tensorpipe/common/cuda.h

```diff
@@ -125,6 +126,15 @@ inline int cudaDeviceForPointer(const void* ptr) {
   return attrs.device;
 }
 
+inline int cudaFixedDeviceForPointer(const CudaLib& cudaLib, const void* ptr) {
+  CUcontext ctx;
+  TP_CUDA_DRIVER_CHECK(cudaLib, cudaLib.ctxGetCurrent(&ctx));
+  TP_CUDA_DRIVER_CHECK(cudaLib, cudaLib.ctxSetCurrent(nullptr));
+  int deviceIdx = cudaDeviceForPointer(ptr);
+  TP_CUDA_DRIVER_CHECK(cudaLib, cudaLib.ctxSetCurrent(ctx));
```

Comment:
Out of curiosity, isn't this effctively equivalent to the following? 
```
cudaCtxPopCurrent(&ctx);
int deviceIdx = cudaDeviceForPointer(ptr);
cudaCtxPushCurrent(ctx);
```

Reply:
Surprisingly it isn't. I can't say why, but that's actually what I tried in the first version of this PR (https://github.com/pytorch/tensorpipe/pull/267/commits/b5ee2b5e4b6f6b8d9111e30b25eb5be8bdade040) and you can see that it caused a deadlock and another failure.